src/drivers/intel/cavs/ssp.c

```diff
@@ -379,68 +349,41 @@ static inline int ssp_set_config(struct dai *dai,
 		goto out;
 	}
 
-	if (F_24000_kHz % config->ssp.mclk_rate == 0) {
-		mdivr_val = F_24000_kHz / config->ssp.mclk_rate;
-	} else {
-		trace_ssp_error("ssp_set_config() error: 24.0MHz / %d Hz MCLK not divisable",
-				config->ssp.mclk_rate);
-		ret = -EINVAL;
-		goto out;
-	}
-
-	if (F_24000_kHz % config->ssp.bclk_rate == 0) {
-		mdiv = F_24000_kHz / config->ssp.bclk_rate;
-	} else {
-		trace_ssp_error("ssp_set_config() error: 24.0MHz / %d Hz BCLK not divisable",
-				config->ssp.bclk_rate);
-		ret = -EINVAL;
-		goto out;
-	}
-#else
-	if (!config->ssp.mclk_rate || config->ssp.mclk_rate > F_24576_kHz) {
-		trace_ssp_error("ssp_set_config() error: "
-				"invalid MCLK = %d Hz (valid < 24567kHz)",
-				config->ssp.mclk_rate);
-		ret = -EINVAL;
-		goto out;
-	}
-	if (!config->ssp.bclk_rate ||
-	    config->ssp.bclk_rate > config->ssp.mclk_rate) {
-		trace_ssp_error("ssp_set_config() error: "
-				"BCLK %d Hz = 0 or > MCLK %d Hz",
-				config->ssp.bclk_rate, config->ssp.mclk_rate);
-		ret = -EINVAL;
-		goto out;
-	}
-	if (F_24576_kHz % config->ssp.mclk_rate == 0) {
+#if defined(PLATFORM_AUDIO_CARDINAL_CLK)
+	if (PLATFORM_AUDIO_CARDINAL_CLK % config->ssp.mclk_rate == 0) {
 		/* select Audio Cardinal clock for MCLK */
 		mdivc |= MCDSS(1);
-		mdivr_val = F_24576_kHz / config->ssp.mclk_rate;
-	} else if (config->ssp.mclk_rate <= F_19200_kHz &&
-		   F_19200_kHz % config->ssp.mclk_rate == 0) {
-		mdivr_val = F_19200_kHz / config->ssp.mclk_rate;
+		mdivr_val = PLATFORM_AUDIO_CARDINAL_CLK /
+			config->ssp.mclk_rate;
+	} else
+#endif
+	if (PLATFORM_XTAL_OSCILLATOR_CLK % config->ssp.mclk_rate == 0) {
```

Comment:
Having an else before an #endif is a very odd style.

Reply:
@plbossart agreed 
I think it combined to be some `else if`

Reply:
Ssp clk defines are moved to platform/clk-map.h. Now I'm searching clock source in ssp_config in for loop.